src/IRI.php

```diff
@@ -10,34 +10,43 @@
 
 /**
  * IRI parser/serialiser/normaliser
+ *
+ * @property ?string $scheme
+ * @property ?string $userinfo
+ * @property ?string $host
+ * @property ?int $port
```

Comment:
For level 3, we would need something like:

```suggestion
 * @property-read int|null $port
 * @property-write int|string|null $port
```

But that is not currently supported https://github.com/phpstan/phpstan/issues/9062